File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.py
import sys
data_temp = sys.stdin if sys.platform == 'linux' else open('입력.txt', 'r')
input_data = data_temp.read().splitlines()

def solution (data) :
  arr = [list(map(int,v.split())) for v in data]
  n,_ = arr.pop(0)
  
  root = {}
  for i in range(1, n+1) :
    root[i] = i 
  
  def find(v) :
    if root[v] == v : return v
    else : 
      root[v] = find(root[v])
      return root[v]
    
  def union(a,b) :
    x,y = find(a), find(b)
    root[y] = x

  total = 0
  arr.sort(key= lambda x : x[2])
  for s,e,w in arr : 
    
    if find(s) == find(e) : continue
    total += w
    union(s,e)
    
  print(total)
  
solution(input_data)

File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.js
const fs = require("fs");
const input = fs
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "입력.txt")
  .toString()
  .trim()
  .split("\n");

function solution(data) {
  const [[n, _], ...units] = data.map((el) => el.split(" ").map(Number));

  const roots = {};
  for (let i = 1; i <= n; i++) {
    roots[i] = i;
  }

  const find = (x) => {
    if (roots[x] !== x) {
      roots[x] = find(roots[x]);
    }
    return roots[x];
  };

  const union = (x, y) => {
    const [a, b] = [find(x), find(y)];
    if (x < y) roots[b] = a;
    else roots[a] = b;
  };

  let total = 0;
  const sorted = units.sort((a, b) => a[2] - b[2]);

  sorted.forEach(([s, e, w]) => {
    if (find(s) !== find(e)) {
      total += w;
      union(s, e);
    }
  });

  console.log(total);
}

solution(input);
